examples/blockchain/chain.rs

```diff
@@ -0,0 +1,206 @@
+        debug!(target: "Blockchain-chain", "Started block creation {:?}", num);
+        // Not extremely random, but good enough.
+        let data: Vec<u8> = (0..size)
+            .map(|i| ((i + i / 999 + (i >> 12)) % 8) as u8)
+            .collect();
+        debug!(target: "Blockchain-chain", "Finished block creation {:?}", num);
+        Block { num, data }
+    }
+}
+
+pub(crate) type BlockPlan = Arc<dyn Fn(BlockNum) -> NodeIndex + Sync + Send + 'static>;
+
+pub(crate) struct ChainConfig {
+    // Our NodeIndex.
+    pub(crate) node_ix: NodeIndex,
+    // Number of random bytes to include in the block.
+    pub(crate) data_size: usize,
+    // Delay between blocks
+    pub(crate) blocktime_ms: u64,
+    // Delay before the first block should be created
+    pub(crate) init_delay_ms: u64,
+    // f(k) means who should author the kth block
+    pub(crate) authorship_plan: BlockPlan,
+}
+
+pub(crate) fn gen_chain_config(
+    node_ix: NodeIndex,
+    n_members: usize,
+    data_size: usize,
+    blocktime_ms: u64,
+    init_delay_ms: u64,
+) -> ChainConfig {
+    //Round robin block authorship plan.
+    let authorship_plan = Arc::new(move |num: u64| NodeIndex((num as usize) % n_members));
+    ChainConfig {
+        node_ix,
+        data_size,
+        blocktime_ms,
+        init_delay_ms,
+        authorship_plan,
+    }
+}
+
+pub(crate) type Data = BlockNum;
+
+struct InnerDataIO {
+    current_block: BlockNum,
+    available_blocks: HashSet<BlockNum>,
+    finalized_tx: UnboundedSender<OrderedBatch<Data>>,
+}
+
+#[derive(Clone)]
+pub(crate) struct DataIO {
+    inner: Arc<Mutex<InnerDataIO>>,
+}
+
+impl aleph_bft::DataIO<Data> for DataIO {
+    type Error = ();
+    fn get_data(&self) -> Data {
+        self.inner.lock().current_block
+    }
+
+    fn check_availability(
+        &self,
+        data: &Data,
+    ) -> Option<Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>>> {
+        if self.inner.lock().available_blocks.contains(data) {
+            return None;
+        }
+        let data_io = self.inner.clone();
+        let data = *data;
+        let fut = async move {
+            loop {
+                if data_io.lock().available_blocks.contains(&data) {
+                    return Ok(());
+                }
+                Delay::new(Duration::from_millis(100)).await;
+            }
+        };
+        Some(Box::pin(fut))
+    }
+
+    fn send_ordered_batch(&mut self, data: OrderedBatch<Data>) -> Result<(), Self::Error> {
+        self.inner
+            .lock()
+            .finalized_tx
+            .unbounded_send(data)
+            .map_err(|_| ())
+    }
+}
+
+impl DataIO {
+    pub(crate) fn new() -> (Self, UnboundedReceiver<OrderedBatch<Data>>) {
+        let (finalized_tx, finalized_rx) = mpsc::unbounded();
+        // We initialize with the "Genesis Block" => number 0.
+        let available_blocks: HashSet<_> = vec![0u64].into_iter().collect();
```

Comment:
No need for the `vec!`, just `iter` over the slice. I'm also quite surprized there is no macro for that.

Reply:
Ok, I did that, but needed to add `cloned()` to make it work as `[1u64, 2u64, 3u64].iter()` iterates over &u64 and not u64 :(

Reply:
I thought there would be a better way of doing this, but even the docs suggest doing it this way... Oh well, good enough.